sda/cmd/api/api.go

```diff
@@ -344,6 +352,118 @@ func deleteFile(c *gin.Context) {
 	c.Status(http.StatusOK)
 }
 
+// reencryptHeader re-encrypts the header of a file using the public key
+// provided in the request header and returns the new header. The function uses
+// gRPC to communicate with the re-encrypt service and handles TLS configuration
+// if needed. The function also handles the case where the CA certificate is
+// provided for secure communication.
+func reencryptHeader(oldHeader []byte, c4ghPubKey string) ([]byte, error) {
+	var opts []grpc.DialOption
+	switch {
+	case Conf.API.Grpc.ClientCreds != nil:
+		opts = append(opts, grpc.WithTransportCredentials(Conf.API.Grpc.ClientCreds))
+	default:
+		opts = append(opts, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	}
+
+	conn, err := grpc.NewClient(fmt.Sprintf("%s:%d", Conf.API.Grpc.Host, Conf.API.Grpc.Port), opts...)
+	if err != nil {
+		log.Errorf("failed to connect to the reencrypt service, reason: %s", err)
+
+		return nil, err
+	}
+	defer conn.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(Conf.API.Grpc.Timeout)*time.Second)
+	defer cancel()
+
+	c := reencrypt.NewReencryptClient(conn)
+	res, err := c.ReencryptHeader(ctx, &reencrypt.ReencryptRequest{Oldheader: oldHeader, Publickey: c4ghPubKey})
+	if err != nil {
+		return nil, err
+	}
+
+	return res.Header, nil
+}
+
+// The downloadFile function download a file re-encrypted with the public key
+// provided in the request header from the inbox. It retrieves the file path
+// from the database using the file ID and user ID.
+func downloadFile(c *gin.Context) {
+	// Get the public key from the request header.
+	c4ghPubKey := c.GetHeader("C4GH-Public-Key")
+
+	pubKey, err := base64.StdEncoding.DecodeString(c4ghPubKey)
+	if err != nil || len(pubKey) == 0 {
+		log.Errorf("bad public key, error: %v", err)
+		c.AbortWithStatusJSON(http.StatusBadRequest, "bad public key")
+
+		return
+	}
+
+	inbox, err := storage.NewBackend(Conf.Inbox)
+	if err != nil {
+		log.Errorf("failed to initialize inbox backend, reason: %v", err)
+		c.AbortWithStatusJSON(http.StatusInternalServerError, "storage backend error")
+
+		return
+	}
+
+	// Retrieve the actual file path for the user's file.
+	fileID := strings.TrimPrefix(c.Param("fileid"), "/")
+	filePath, err := Conf.API.DB.GetInboxFilePathFromID(
+		strings.TrimPrefix(c.Param("username"), "/"),
+		fileID,
+	)
+	if err != nil {
+		log.Errorf("getting file path from fileID (%s) failed, reason: %v", fileID, err)
+		c.AbortWithStatusJSON(http.StatusNotFound, "failed to retrieve inbox file path")
+
+		return
+	}
+
+	// Get inbox file handle
+	file, err := inbox.NewFileReader(filePath)
+	if err != nil {
+		log.Errorf("inbox file %s not found or failed to read, %s", filePath, err)
```

Comment:
```suggestion
		log.Errorf("inbox file %s not found or failed to read, %s", filePath, err.Error())
```

Since you use `%s` you need to use the string portion of the error.